net/Socket.cpp

```diff
@@ -788,53 +790,39 @@ void SocketPoll::closeAllSockets()
     assert(_newSockets.size() == 0);
 }
 
-void SocketPoll::takeSocket(const std::shared_ptr<SocketPoll> &fromPoll,
-                            const std::shared_ptr<Socket> &inSocket)
+void SocketPoll::takeSocket(const std::shared_ptr<SocketPoll>& fromPoll,
+                            const std::shared_ptr<SocketPoll>& toPoll,
+                            const std::shared_ptr<Socket>& inSocket)
 {
     std::mutex mut;
     std::condition_variable cond;
     bool transferred = false;
 
     // Important we're not blocking the fromPoll thread.
-    ASSERT_CORRECT_THREAD();
+    toPoll->assertCorrectThread(__FILE__, __LINE__);
 
-    // hold a reference during transfer
-    std::shared_ptr<Socket> socket = inSocket;
+    int socketFD = inSocket->getFD();
 
-    SocketPoll *toPoll = this;
-    fromPoll->addCallback([fromPoll,socket,&mut,&cond,&transferred,toPoll](){
-        auto it = std::find(fromPoll->_pollSockets.begin(),
-                            fromPoll->_pollSockets.end(), socket);
-        if (it != fromPoll->_pollSockets.end())
-        {
-            // Erasing messes up the tracking of poll results in 'poll'
-            // leave to be added to toErase and cleaned later.
-            *it = nullptr;
-        }
-        else
-            LOG_WRN("Trying to move socket out of the wrong poll");
-
-        // sockets in transit are un-owned
-        SocketThreadOwnerChange::resetThreadOwner(*socket);
-
-        toPoll->insertNewSocket(socket);
+    fromPoll->transferSocketTo(inSocket, toPoll,
+        [](const std::shared_ptr<Socket>& /*moveSocket*/){},
```

Comment:
Ok - slightly counter-intuitively this blocks until the socket is removed from the poll - prolly it should have a comment that says so - good to have the assertCorrectThread to check and avoid the deadlock =)